compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java

```diff
@@ -1360,22 +1363,81 @@ private boolean isValidAnnotationType(BType type) {
         }
 
         switch (type.tag) {
+            case TypeTags.MAP:
+                BType constraintType = ((BMapType) type).constraint;
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(constraintType);
             case TypeTags.RECORD:
-//                BRecordType recordType = (BRecordType) type;
-//                return recordType.fields.stream().allMatch(field -> types.isAnydata(field.type)) &&
-//                        (recordType.sealed || types.isAnydata(recordType.restFieldType));
+                BRecordType recordType = (BRecordType) type;
+                for (BField field : recordType.fields.values()) {
+                    if (!isAnyDataOrReadOnlyTypeSkippingObjectType(field.type)) {
+                        return false;
+                    }
+                }
+
+                BType recordRestType = recordType.restFieldType;
+                if (recordRestType == null || recordRestType == symTable.noType) {
+                    return true;
+                }
+
+                return isAnyDataOrReadOnlyTypeSkippingObjectType(recordRestType);
+            case TypeTags.ARRAY:
+                return isValidAnnotationType(((BArrayType) type).eType);
```

Comment:
We need to keep the part of the original check where we first check if the member type is a map or a record (https://github.com/ballerina-platform/ballerina-lang/blob/master/compiler/ballerina-lang/src/main/java/org/wso2/ballerinalang/compiler/semantics/analyzer/SymbolEnter.java#L1372) or we will be allowing an array of `true` also.

We should ideally have a negative test for the same.

Reply:
Fixed : 0d31fb64b78458d94895d4c1fdbb91c53688f09d